retrofit2/src/main/java/com/linecorp/armeria/retrofit2/ArmeriaCallFactory.java

```diff
@@ -0,0 +1,200 @@
+import okhttp3.ResponseBody;
+import okio.Buffer;
+
+public class ArmeriaCallFactory implements Call.Factory {
+
+    private final HttpClient httpClient;
+
+    public ArmeriaCallFactory(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public Call newCall(Request request) {
+        return new ArmeriaCall(this, request);
+    }
+
+    private static class ArmeriaCall implements Call {
+
+        private final ArmeriaCallFactory callFactory;
+        private final Request request;
+
+        // Guarded by this.
+        private boolean executed;
+
+        volatile boolean canceled;
+
+        private CompletableFuture<AggregatedHttpMessage> future;
+
+        ArmeriaCall(ArmeriaCallFactory callFactory, Request request) {
+            this.callFactory = callFactory;
+            this.request = request;
+        }
+
+        private HttpResponse doCall(HttpClient httpClient, Request request) {
+            URL url = request.url().url();
+            String uri = url.getPath();
+            if (url.getQuery() != null) {
+                uri += '?' + url.getQuery();
+            }
+            final HttpHeaders headers;
+            switch (request.method()) {
+                case "GET":
+                    headers = HttpHeaders.of(HttpMethod.GET, uri);
+                    break;
+                case "HEAD":
+                    headers = HttpHeaders.of(HttpMethod.HEAD, uri);
+                    break;
+                case "POST":
+                    headers = HttpHeaders.of(HttpMethod.POST, uri);
+                    break;
+                case "DELETE":
+                    headers = HttpHeaders.of(HttpMethod.DELETE, uri);
+                    break;
+                case "PUT":
+                    headers = HttpHeaders.of(HttpMethod.PUT, uri);
+                    break;
+                case "PATCH":
+                    headers = HttpHeaders.of(HttpMethod.PATCH, uri);
+                    break;
+                case "OPTIONS":
+                    headers = HttpHeaders.of(HttpMethod.OPTIONS, uri);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Invalid HTTP method:" + request.method());
+            }
+            request.headers().toMultimap().forEach(
+                    (key, values) -> headers.add(HttpHeaderNames.of(key), values));
+            if (request.body() != null) {
+                headers.set(HttpHeaderNames.CONTENT_TYPE, request.body().contentType().toString());
+                Buffer contentBuffer = new Buffer();
+                try {
+                    request.body().writeTo(contentBuffer);
+                    return httpClient.execute(headers, contentBuffer.readByteArray());
+                } catch (IOException e) {
+                    throw new IllegalArgumentException(
+                            "Failed to convert RequestBody to SimpleRequest.content." + request.method(), e);
```

Comment:
- How about move the `httpClient.execute()` call out of this try block so that it's really sure that the exception is because of conversion failure?
- `SimpleRequest` -> `HttpData` ?